### PR TITLE
Attentionpooling ecapa

### PR DESCRIPTION
**Pull request by nithinraok**
* Adds new modified ecapa ecoder for better speaker embeddings
* adds attention pooling decoder 
* adds TAP pooling 
* tests

### CODE REVIEW

File: examples/speaker_recognition/conf/SpeakerNet_ECAPA.yaml [hunk 1]
```diff
@@ -0,0 +1,94 @@
+name: &name "SpeakerNet"
+sample_rate: &sample_rate 16000
```
Comment (titu1994) [inline examples/speaker_recognition/conf/SpeakerNet_ECAPA.yaml:2]: All of these values after name should be inside "model". Remove "&xyz" and use hydra interpolation

File: examples/speaker_recognition/conf/SpeakerNet_ECAPA.yaml [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+  decoder:
+    _target_: nemo.collections.asr.modules.SpeakerDecoder
+    feat_in: &enc_feat_out 3072
+    num_classes: 7205
+    pool_mode: 'attention'
+    emb_sizes: 192
+    angular: True
+
+  loss:
```
Comment (titu1994) [inline examples/speaker_recognition/conf/SpeakerNet_ECAPA.yaml:61]: (Maybe) - Use `_target_` and instantiate loss so that is it customizable.

File: nemo/collections/asr/modules/conv_asr.py [hunk 3]
```diff
@@ -342,6 +342,77 @@ def num_classes(self):
         return self._num_classes
 
 
+class ECAPA_Encoder(NeuralModule, Exportable):
+    """
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:346]: No underscore - use `ECAPAEncoder`.

File: nemo/collections/asr/modules/conv_asr.py [hunk 4]
```diff
@@ -342,6 +342,77 @@ def num_classes(self):
...
+    Modified ECAPA Encoder layer without Res2Net module for faster training and inference which achieves
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:347]: Post link to Res2Net paper.

File: nemo/collections/asr/modules/conv_asr.py [hunk 5]
```diff
@@ -20,12 +20,12 @@
...
-    JasperBlock,
-    MaskedConv1d,
+from nemo.collections.asr.parts.submodules.jasper import JasperBlock, MaskedConv1d, init_weights, jasper_activations
+from nemo.collections.asr.parts.submodules.tdnn_attention import (
+    AttentivePoolLayer,
+    SE_TDNN_Module,
     StatsPoolLayer,
-    init_weights,
-    jasper_activations,
+    TDNN_Module,
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:28]: No _. Use TDNNModule

File: nemo/collections/asr/modules/conv_asr.py [hunk 6]
```diff
@@ -342,6 +342,77 @@ def num_classes(self):
...
+        """Returns definitions of module output ports.
+        """
+        return OrderedDict(
+            {
+                "outputs": NeuralType(('B', 'D', 'T'), AcousticEncodedRepresentation()),
+                "encoded_lengths": NeuralType(tuple('B'), LengthsType()),
+            }
+        )
+
+    def __init__(self, feat_in, filters, kernel_sizes, dilations, scale=8, init_mode='xavier_uniform'):
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:384]: Add type info for init

File: nemo/collections/asr/modules/conv_asr.py [hunk 7]
```diff
@@ -342,6 +342,77 @@ def num_classes(self):
...
+        super(ECAPA_Encoder, self).__init__()
+        self.layers = nn.ModuleList()
+        self.layers.append(TDNN_Module(feat_in, filters[0], kernel_size=kernel_sizes[0], dilation=dilations[0]))
+
+        for i in range(len(filters) - 2):
+            self.layers.append(
+                SE_TDNN_Module(
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:391]: Rename to TDNNSEModule

File: nemo/collections/asr/modules/conv_asr.py [hunk 8]
```diff
@@ -351,7 +422,7 @@ class SpeakerDecoder(NeuralModule, Exportable):
         num_classes (int): Number of unique speakers in dataset
         emb_sizes (list) : shapes of intermediate embedding layers (we consider speaker embbeddings from 1st of this layers)
                 Defaults to [1024,1024]
-        pool_mode (str) : Pooling stratergy type. options are 'gram','xvector','superVector'.
+        pool_mode (str) : Pooling stratergy type. options are 'xvector','tap', 'ecapa'
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:425]: Are the older types being deprecated? Raise error explicitly for old values then

File: nemo/collections/asr/modules/conv_asr.py [hunk 9]
```diff
@@ -382,35 +458,37 @@ def output_types(self):
         )
 
     def __init__(
-        self, feat_in, num_classes, emb_sizes=None, pool_mode='xvector', angular=False, init_mode="xavier_uniform",
+        self,
+        feat_in,
+        num_classes,
+        emb_sizes=[256],
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:464]: Do not use [256] in function. Use None, and check below if None then set emb_size = [256].

File: nemo/collections/asr/modules/conv_asr.py [hunk 10]
```diff
@@ -382,35 +458,37 @@ def output_types(self):
         )
 
     def __init__(
-        self, feat_in, num_classes, emb_sizes=None, pool_mode='xvector', angular=False, init_mode="xavier_uniform",
+        self,
+        feat_in,
+        num_classes,
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:463]: Add types to all new methods and classes

File: nemo/collections/asr/modules/conv_asr.py [hunk 11]
```diff
@@ -382,35 +458,37 @@ def output_types(self):
...
+        emb_sizes=[256],
+        pool_mode='xvector',
+        angular=False,
+        attention_channels=128,
+        init_mode="xavier_uniform",
     ):
-        super().__init__()
+        super(SpeakerDecoder, self).__init__()
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:470]: Revert. `super().__init__()` is the proper standard for Py3

File: nemo/collections/asr/modules/conv_asr.py [hunk 12]
```diff
@@ -382,35 +458,37 @@ def output_types(self):
...
-            emb_sizes = [512, 512]
-
-        self.input_feat_in = feat_in
         self._num_classes = num_classes
-        self._pooling = StatsPoolLayer(feat_in=feat_in, pool_mode=pool_mode)
-        self._feat_in = self._pooling.feat_in
-
-        shapes = [self._feat_in]
+        self.pool_mode = pool_mode.lower()
+        if self.pool_mode == 'xvector' or self.pool_mode == 'tap':
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:478]: Provide documentation above what `tap` means, same for below types

File: nemo/collections/asr/modules/conv_asr.py [hunk 13]
```diff
@@ -419,32 +497,42 @@ def __init__(
 
         self.apply(lambda x: init_weights(x, mode=init_mode))
 
-    def affineLayer(self, inp_shape, out_shape, learn_mean=True):
-        layer = nn.Sequential(
-            nn.Linear(inp_shape, out_shape),
-            nn.BatchNorm1d(out_shape, affine=learn_mean, track_running_stats=True),
-            nn.ReLU(),
-        )
+    def affineLayer(
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:500]: Snake case only. `affine_layer`

File: nemo/collections/asr/modules/conv_asr.py [hunk 14]
```diff
@@ -370,7 +441,12 @@ def input_example(self):
 
     @property
     def input_types(self):
-        return OrderedDict({"encoder_output": NeuralType(('B', 'D', 'T'), AcousticEncodedRepresentation())})
+        return OrderedDict(
+            {
+                "encoder_output": NeuralType(('B', 'D', 'T'), AcousticEncodedRepresentation()),
+                "length": NeuralType(('B',), LengthsType()),
+            }
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:448]: Make the neural type optional since below lengths=None in the forward call.

File: nemo/collections/asr/parts/submodules/tdnn_attention.py [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+    Statistics and time average pooling (TAP) layer
+    This computes mean and variance statistics across time dimension (dim=-1)
+    input:
+        feat_in: input channel feature length
+    output:
+        pooled: statistics of feature input
+    """
+
+    def __init__(self, feat_in, pool_mode='xvector'):
+        super().__init__()
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/tdnn_attention.py:34]: Provide info of all supported pool_mode

File: nemo/collections/asr/parts/submodules/tdnn_attention.py [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+    output:
+        mean: mean of input features
+        std: stadard deviation of input features
+    """
+    mean = torch.sum((m * x), dim=dim)
+    std = torch.sqrt((m * (x - mean.unsqueeze(dim)).pow(2)).sum(dim).clamp(eps))
+    return mean, std
+
+
+class TDNN_Module(nn.Module):
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/tdnn_attention.py:88]: rename class to TDNNModule.

File: nemo/collections/asr/parts/submodules/tdnn_attention.py [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+        else:
+            max_len = input.size(2)
+            mask, num_values = lens_to_mask(length, max_len=max_len, device=input.device)
+            x = torch.sum((input * mask), dim=2, keepdim=True) / (num_values)
+
+        out = self.se_layer(x)
+        return out * input
+
+
+class SE_TDNN_Module(nn.Module):
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/tdnn_attention.py:160]: Rename to TDNNSEModule.

File: nemo/collections/asr/parts/submodules/tdnn_attention.py [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+        self.activation = nn.ReLU()
+        self.bn = nn.BatchNorm1d(out_filters)
+
+    def forward(self, x, length=None):
+        x = self.conv_layer(x)
+        x = self.activation(x)
+        return self.bn(x)
+
+
+class SE_Module(nn.Module):
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/tdnn_attention.py:124]: Why reimplement SEBlock from jasper.py?? Revert this and use that. If there is really no way to use that, only then rename this to `TDNNSEBlock` and use it in `TDNNSEModule`

File: tests/collections/asr/test_speaker_label_models.py [hunk 19]
```diff
@@ -61,3 +61,34 @@ def test_constructor(self):
...
         self.assertTrue(isinstance(instance2, EncDecSpeakerLabelModel))
+
+    @pytest.mark.unit
+    def test_ecapa_enc_dec(self):
+        preprocessor = {'cls': 'nemo.collections.asr.modules.AudioToMelSpectrogramPreprocessor', 'params': dict({})}
+        encoder = {
+            'cls': 'nemo.collections.asr.modules.ECAPA_Encoder',
+            'params': {
+                'feat_in': 80,
+                'filters': [1024, 1024, 1024, 1024, 3072],
```
Comment (titu1994) [inline tests/collections/asr/test_speaker_label_models.py:72]: Dont use such large filters for tests, it will slow it down.

**Review by titu1994 — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/modules/conv_asr.py (hunk 8 above)
Comment (nithinraok) [reply nemo/collections/asr/modules/conv_asr.py:425]: older poolings were added but we didn't support them. They were just there for experiment and internal research purposes when I was first designing SpeakerNet

File: nemo/collections/asr/modules/conv_asr.py (hunk 9 above)
Comment (nithinraok) [reply nemo/collections/asr/modules/conv_asr.py:464]: using int

File: nemo/collections/asr/parts/submodules/tdnn_attention.py (hunk 18 above)
Comment (nithinraok) [reply nemo/collections/asr/parts/submodules/tdnn_attention.py:124]: `jasper.py` uses linear layer and also order is different. Observed conv1d layers with kernel size=1 gave better numbers and also converged faster

File: tests/collections/asr/test_speaker_label_models.py (hunk 19 above)
Comment (nithinraok) [reply tests/collections/asr/test_speaker_label_models.py:72]: you are right

File: nemo/collections/asr/parts/submodules/tdnn_attention.py (hunk 18 above)
Comment (titu1994) [reply nemo/collections/asr/parts/submodules/tdnn_attention.py:124]: What's different about order ? And conv 1d with pointwise kernel is mathematically equivalent to dense layer

Comment (titu1994) [reply nemo/collections/asr/parts/submodules/tdnn_attention.py:124]: They do have different initializations though. Hmm fine just rename it for now

Comment (nithinraok) [reply nemo/collections/asr/parts/submodules/tdnn_attention.py]: it doesn't; perform batch norm and also I take care of length masking

File: nemo/collections/asr/parts/submodules/tdnn_attention.py [hunk 20]
```diff
@@ -0,0 +1,279 @@
...
+        self.activation = nn.ReLU()
+        self.bn = nn.BatchNorm1d(out_filters)
+
+    def forward(self, x, length=None):
+        x = self.conv_layer(x)
+        x = self.activation(x)
+        return self.bn(x)
+
+
+class SEModule(nn.Module):
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/tdnn_attention.py:137]: Rename to MaskedSE or something similar since this implementation deviates from the SE block in computer vision significantly

**Review by titu1994 — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/parts/submodules/tdnn_attention.py (hunk 20 above)
Comment (nithinraok) [reply nemo/collections/asr/parts/submodules/tdnn_attention.py]: Done

**Review by titu1994 — APPROVED** [review]
